File: java/src/main/java/lab/gabon/service/TaskService.java
package lab.gabon.service;

import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.temporal.IsoFields;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import lab.gabon.common.AppError;
import lab.gabon.common.AppException;
import lab.gabon.model.entity.TaskDefinition;
import lab.gabon.model.entity.TaskProgress;
import lab.gabon.model.response.TaskResponses.TaskItemResponse;
import lab.gabon.repository.CustomerRepository;
import lab.gabon.repository.TaskDefinitionRepository;
import lab.gabon.repository.TaskProgressRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class TaskService {

  private static final ZoneId SHANGHAI_ZONE = ZoneId.of("Asia/Shanghai");

  private static final short TASK_TYPE_DAILY = 1;
  private static final short TASK_TYPE_WEEKLY = 2;
  private static final short TASK_TYPE_MONTHLY = 3;

  private static final short STATUS_COMPLETED = 2;
  private static final short STATUS_CLAIMED = 3;

  private final TaskDefinitionRepository taskDefRepo;
  private final TaskProgressRepository taskProgressRepo;
  private final CustomerRepository customerRepo;

  public TaskService(
      TaskDefinitionRepository taskDefRepo,
      TaskProgressRepository taskProgressRepo,
      CustomerRepository customerRepo) {
    this.taskDefRepo = taskDefRepo;
    this.taskProgressRepo = taskProgressRepo;
    this.customerRepo = customerRepo;
  }

  public List<TaskItemResponse> listTasks(long customerId) {
    var now = Instant.now();
    var definitions = taskDefRepo.findActiveDefinitions();

    // Upsert progress for each definition with its matching period key
    for (var def : definitions) {
      var periodKey = generatePeriodKey(def.taskType(), now);
      taskProgressRepo.upsertProgress(
          customerId, def.id(), def.targetCount(), periodKey, def.rewardDiamonds());
    }

    // Collect progress rows grouped by period key
    var taskTypes = definitions.stream().map(TaskDefinition::taskType).distinct().toList();

    var progressByTaskId = new HashMap<Long, TaskProgress>();
    for (var tt : taskTypes) {
      var periodKey = generatePeriodKey(tt, now);
      var rows = taskProgressRepo.findByCustomerAndPeriod(customerId, periodKey);
      for (var row : rows) {
        progressByTaskId.put(row.taskId(), row);
      }
    }

    // Merge definitions with progress
    var items = new ArrayList<TaskItemResponse>(definitions.size());
    for (var def : definitions) {
      var periodKey = generatePeriodKey(def.taskType(), now);
      var progress = progressByTaskId.get(def.id());

      items.add(
          new TaskItemResponse(
              def.id(),
              def.taskCode(),
              def.taskName(),
              def.description(),
              def.targetCount(),
              progress != null ? progress.currentCount() : 0,
              def.rewardDiamonds(),
              progress != null ? progress.taskStatus() : 1,
              periodKey));
    }

    return items;
  }

  @Transactional
  public int claimReward(long customerId, long progressId) {
    var progress =
        taskProgressRepo
            .findByIdForUpdate(progressId, customerId)
            .orElseThrow(() -> new AppException(new AppError.TaskNotClaimable()));

    if (progress.taskStatus() != STATUS_COMPLETED) {
      throw new AppException(new AppError.TaskNotClaimable());
    }

    taskProgressRepo.claimReward(progressId);
    customerRepo.updateDiamondBalance(customerId, progress.rewardDiamonds());

    return progress.rewardDiamonds();
  }

  static String generatePeriodKey(short taskType, Instant instant) {
    ZonedDateTime zdt = instant.atZone(SHANGHAI_ZONE);
    return switch (taskType) {
      case TASK_TYPE_DAILY -> zdt.toLocalDate().toString();
      case TASK_TYPE_WEEKLY -> {
        int year = zdt.get(IsoFields.WEEK_BASED_YEAR);
        int week = zdt.get(IsoFields.WEEK_OF_WEEK_BASED_YEAR);
        yield "%d-W%02d".formatted(year, week);
      }
      case TASK_TYPE_MONTHLY -> "%d-%02d".formatted(zdt.getYear(), zdt.getMonthValue());
      default -> throw new IllegalArgumentException("Unknown task type: " + taskType);
    };
  }
}

File: kotlin/src/main/kotlin/lab/gabon/service/TaskService.kt
package lab.gabon.service

import lab.gabon.model.TaskType
import lab.gabon.repository.SignInRepo
import lab.gabon.repository.TaskRepo
import lab.gabon.repository.TaskWithProgressRow
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.temporal.IsoFields

data class TaskItem(
    val taskId: Long,
    val taskCode: String,
    val taskName: String,
    val taskType: Short,
    val targetCount: Int,
    val rewardDiamonds: Int,
    val progressId: Long,
    val currentCount: Int,
    val taskStatus: Short,
)

data class SignInResult(
    val diamonds: Int,
    val periodKey: String,
)

class TaskService(
    private val taskRepo: TaskRepo,
    private val signInRepo: SignInRepo,
) {
    companion object {
        val SHANGHAI_ZONE: ZoneId = ZoneId.of("Asia/Shanghai")
        const val SIGN_IN_REWARD = 1
    }

    suspend fun listTasks(
        customerId: Long,
        taskType: Short? = null,
        taskStatus: Short? = null,
        now: Instant = Instant.now(),
    ): List<TaskItem> {
        val definitions = taskRepo.findActiveTaskDefinitions(taskType)

        // Upsert progress for each definition with its matching period key
        for (def in definitions) {
            val periodKey = generatePeriodKey(def.taskType, now)
            taskRepo.upsertProgress(
                customerId = customerId,
                taskId = def.id,
                periodKey = periodKey,
                targetCount = def.targetCount,
                rewardDiamonds = def.rewardDiamonds,
            )
        }

        // Collect all period keys for the query (daily, weekly, monthly)
        // We query for each task type's period key
        val results = mutableListOf<TaskWithProgressRow>()
        val taskTypes = definitions.map { it.taskType }.distinct()

        for (tt in taskTypes) {
            val periodKey = generatePeriodKey(tt, now)
            val filtered =
                taskRepo.listProgressWithTasks(
                    customerId = customerId,
                    periodKey = periodKey,
                    taskType = tt,
                    taskStatus = taskStatus,
                )
            results.addAll(filtered)
        }

        return results.map { it.toTaskItem() }
    }

    suspend fun claimReward(
        customerId: Long,
        progressId: Long,
    ): Int = taskRepo.claimReward(progressId, customerId)

    suspend fun signIn(
        customerId: Long,
        now: Instant = Instant.now(),
    ): SignInResult {
        val periodKey = generatePeriodKey(TaskType.DAILY.value, now)
        signInRepo.signIn(customerId, periodKey, SIGN_IN_REWARD)
        return SignInResult(diamonds = SIGN_IN_REWARD, periodKey = periodKey)
    }

    suspend fun incrementWatchProgress(
        customerId: Long,
        now: Instant = Instant.now(),
    ) {
        // Find the daily watch task (category=1) progress for today
        val periodKey = generatePeriodKey(TaskType.DAILY.value, now)
        val definitions = taskRepo.findActiveTaskDefinitions(TaskType.DAILY.value)
        val watchTask = definitions.find { it.taskCategory.toInt() == 1 } ?: return

        val progress =
            taskRepo.findProgressByCustomerAndTask(
                customerId = customerId,
                taskId = watchTask.id,
                periodKey = periodKey,
            ) ?: return

        if (progress.taskStatus.toInt() == 1) { // IN_PROGRESS
            taskRepo.incrementProgress(progress.id)
        }
    }

    fun generatePeriodKey(
        taskType: Short,
        instant: Instant,
    ): String {
        val zdt: ZonedDateTime = instant.atZone(SHANGHAI_ZONE)
        return when (taskType) {
            TaskType.DAILY.value -> zdt.toLocalDate().toString() // "2026-03-19"
            TaskType.WEEKLY.value -> {
                val year = zdt.get(IsoFields.WEEK_BASED_YEAR)
                val week = zdt.get(IsoFields.WEEK_OF_WEEK_BASED_YEAR)
                "%d-W%02d".format(year, week) // "2026-W12"
            }
            TaskType.MONTHLY.value -> "%d-%02d".format(zdt.year, zdt.monthValue) // "2026-03"
            else -> throw IllegalArgumentException("Unknown task type: $taskType")
        }
    }

    private fun TaskWithProgressRow.toTaskItem(): TaskItem =
        TaskItem(
            taskId = taskId,
            taskCode = taskCode,
            taskName = taskName,
            taskType = taskType,
            targetCount = targetCount,
            rewardDiamonds = rewardDiamonds,
            progressId = progressId,
            currentCount = currentCount,
            taskStatus = taskStatus,
        )
}
